goldmane/pkg/emitter/emitter.go

```diff
@@ -0,0 +1,264 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package emitter
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/projectcalico/calico/goldmane/pkg/aggregator"
+)
+
+var (
+	maxRetries   = 15
+	configMapKey = types.NamespacedName{Name: "flow-emitter-state", Namespace: "calico-system"}
+)
+
+// Emitter is a type that emits aggregated Flow objects to an HTTP endpoint.
+type Emitter struct {
+	client *emitterClient
+
+	kcli client.Client
+
+	// Configuration for emitter endpoint.
+	url        string
+	caCert     string
+	clientKey  string
+	clientCert string
+	serverName string
+
+	// Use a rate limited workqueue to manage bucket emission.
+	buckets *bucketCache
+	q       workqueue.TypedRateLimitingInterface[bucketKey]
+
+	// Track the latest timestamp of emitted flows. This helps us avoid emitting the same flow multiple times
+	// on restart.
+	latestTimestamp int64
+}
+
+// Make sure Emitter implements the Receiver interface to be able to receive aggregated Flows.
+var _ aggregator.Sink = &Emitter{}
+
+func NewEmitter(opts ...Option) *Emitter {
+	e := &Emitter{
+		buckets: newBucketCache(),
+		q: workqueue.NewTypedRateLimitingQueue(
+			workqueue.NewTypedMaxOfRateLimiter(
+				workqueue.NewTypedItemExponentialFailureRateLimiter[bucketKey](1*time.Second, 30*time.Second),
+				&workqueue.TypedBucketRateLimiter[bucketKey]{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+			)),
+	}
+
+	for _, opt := range opts {
+		opt(e)
+	}
+
+	var err error
+	e.client, err = newEmitterClient(e.url, e.caCert, e.clientKey, e.clientCert, e.serverName)
+	if err != nil {
+		logrus.Fatalf("Error creating emitter client: %v", err)
+	}
+	logrus.WithField("url", e.url).Info("Created emitter client.")
+
+	if e.kcli == nil {
+		logrus.Warn("No k8s client provided, will not be able to cache state.")
+	}
+
+	return e
+}
+
+func (e *Emitter) Run(stopCh chan struct{}) {
+	// Start by loading any state cached in our configmap, which will allow us to better pick up where we left off
+	// in the event of a restart.
+	if err := e.loadCachedState(); err != nil {
+		logrus.Errorf("Error loading cached state: %v", err)
+	}
+
+	// This is the main loop for the emitter. It listens for new batches of flows to emit and emits them.
+	for {
+		// Get pending work from the queue.
+		key, quit := e.q.Get()
+		if quit {
+			return
+		}
+		e.q.Done(key)
+
+		// Emit the bucket.
+		bucket, _ := e.buckets.get(key)
```

Comment:
Should we handle if key does not exists? I'm not sure if that's possible, but it feels we should even if it's rare case or maybe if it's programmer mistake (which it should panic). WDYT?

Reply:
Yeah, definitely should consider this case. I _think_ it can only happen with programming error? But I'll log an error and handle it to be safe.